ENH: spatial: add `axis` optional argument for `distance.jensenshannon`.

AtsushiSakai opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
fix #11325

#### What does this implement/fix?
<!--Please explain your changes.-->
I added `axis` optional argument for `spatial.distance.jensenshannon`.
I added tests for it and update docstrings.

tupui commented on scipy/spatial/distance.py:
```diff
@@ -1211,7 +1211,7 @@ def canberra(u, v, w=None):
+def jensenshannon(p, q, base=None, axis=0):
     """
     Compute the Jensen-Shannon distance (metric) between
     two 1-D probability arrays. This is the square root
```
Need to be updated probably.

AtsushiSakai replied:
Thanks. I updated.

tupui commented on scipy/spatial/tests/test_distance.py:
```diff
@@ -2200,3 +2200,28 @@ def test_yule_all_same():
+                  [17, 18, 19, 20],
+                  [21, 22, 23, 24]])
+
+    assert_almost_equal(jensenshannon(a, b, axis=0),
+                        [0.1954288, 0.1447697, 0.1138377, 0.0927636])
+    assert_almost_equal(jensenshannon(a, b, axis=1),
+                        [0.1402339, 0.0399106, 0.0201815])
```
I checked all these, it's correct 👍

tupui commented on scipy/spatial/distance.py:
```diff
@@ -1256,16 +1259,20 @@ def jensenshannon(p, q, base=None):
 
     """
     p = np.asarray(p)
     q = np.asarray(q)
```
I would add a validation of the shapes here. Now it's more complex with the axis and could help the user.

peterbell10 replied:
Not sure that's a good idea. NumPy will give an appropriate error on shape mismatch and reimplementing the correct checks here, keeping in mind broadcasting would be fairly complicated/another source of bugs.

tupui replied:
Ok closing this then.

tupui commented on scipy/spatial/distance.py:
```diff
@@ -1239,6 +1239,9 @@ def jensenshannon(p, q, base=None):
         scipy.stats.entropy.
+    axis : int, optional
+        Axis along which the Jensen-Shannon distances are computed. The default
+        is 0.
 
```
```suggestion
        is 0.

        .. versionadded:: 1.7.0
```

tupui commented on scipy/spatial/distance.py:
```diff
@@ -1239,6 +1239,9 @@ def jensenshannon(p, q, base=None):
+        is 0.
 
     Returns
     -------
```
The return type is not just `double` anymore.

tupui commented on scipy/spatial/distance.py:
```diff
@@ -1256,16 +1259,20 @@ def jensenshannon(p, q, base=None):
     0.46450140402245893
     >>> distance.jensenshannon([1.0, 0.0, 0.0], [1.0, 0.0, 0.0])
     0.0
+    >>> distance.jensenshannon([[1.0, 2.0]], [[0.5, 1.5]], axis=0)
```
Maybe having a larger example would be more explanatory. I was puzzled at first with this first example.

tupui reviewed (changes requested):
Thanks, good addition making this function more versatile. I just noted a few minor things.

peterbell10 commented on scipy/spatial/distance.py:
```diff
@@ -1211,7 +1211,7 @@ def canberra(u, v, w=None):
 
 
-def jensenshannon(p, q, base=None):
+def jensenshannon(p, q, base=None, axis=0):
```
```suggestion
def jensenshannon(p, q, base=None, *, axis=0, keepdim=False):
```

AtsushiSakai replied:
I'm not sure what the `*` is for?

tupui replied:
This is to say that in order to use the parameters after it, you must call them explicitly `axis=0`, you cannot pass just `0`.

tupui replied:
I am not sure about the need for `keepdim`. @peterbell10 ?

If we add it, we will need to add `.. versionadded:: 1.7.0` here too.

AtsushiSakai replied:
Thanks @tupui I didn't know that.

> If we add it, we will need to add .. versionadded:: 1.7.0 here too.
I added it.

AtsushiSakai commented:
I think I addressed all review comments. PTAL again.

Also, I realized current doc has doc rendering issue, so I fixed it by adding note section.
Before:
<img width="749" alt="SS 2021-05-01 at 13 07 59" src="https://user-images.githubusercontent.com/3813847/116774379-4804f480-aa97-11eb-98b4-dc9a1e5a8d3a.png">
After:
<img width="785" alt="SS 2021-05-01 at 14 24 05" src="https://user-images.githubusercontent.com/3813847/116774385-4e936c00-aa97-11eb-926c-fee6ec498dbf.png">

tupui commented on scipy/spatial/distance.py:
```diff
@@ -1239,11 +1239,25 @@ def jensenshannon(p, q, base=None):
+    keepdims : bool, optional
+        If this is set to `True`, the reduced axes are left in the
+        result as dimensions with size one. With this option,
+        the result will broadcast correctly against the input array.
```
```suggestion
        the result will broadcast correctly against the input array.
        Default is False.
```

tupui reviewed (approved):
Just a minor comment. LGTM otherwise. I will wait for the clarification about `keepdim` from @peterbell10 and I will merge.

tupui commented:
I went to read up about this `keepdim`. I am neutral about it. If this can help anybody why not. This way we keep the structure intact. So merging.

Thanks @AtsushiSakai!
